storage_controller/src/tenant_shard.rs

```diff
@@ -740,90 +733,261 @@ impl TenantShard {
+            // modified to decrease N.
+
+            let mut secondary_scores = self
+                .intent
+                .get_secondary()
+                .iter()
+                .map(|node_id| {
+                    (
+                        *node_id,
+                        scheduler.compute_node_score::<NodeSecondarySchedulingScore>(
+                            *node_id,
+                            &self.preferred_az_id,
+                            &schedule_context,
+                        ),
+                    )
+                })
+                .collect::<Vec<_>>();
+
+            if secondary_scores.iter().any(|score| score.1.is_none()) {
+                // Don't have full list of scores, so can't make a good decision about which to drop unless
+                // there is an obvious one in the wrong AZ
+                for secondary in self.intent.get_secondary() {
+                    if scheduler.get_node_az(secondary) == self.preferred_az_id {
+                        return Some(ScheduleOptimization {
+                            sequence: self.sequence,
+                            action: ScheduleOptimizationAction::RemoveSecondary(*secondary),
+                        });
+                    }
                 }
+
+                // Fall through: we didn't identify one to remove.  This ought to be rare.
+                tracing::warn!("Keeping extra secondaries: can't determine which of {:?} to remove (some nodes offline?)",
+                self.intent.get_secondary()
+            );
             } else {
+                secondary_scores.sort_by_key(|score| score.1.unwrap());
+                let victim = secondary_scores.last().unwrap().0;
+                return Some(ScheduleOptimization {
+                    sequence: self.sequence,
+                    action: ScheduleOptimizationAction::RemoveSecondary(victim),
+                });
+            }
+        }
+
+        let replacement = self.find_better_location::<AttachedShardTag>(
+            scheduler,
+            &schedule_context,
+            attached,
+            &[], // Don't exclude secondaries: our preferred attachment location may be a secondary
+        );
+
+        // We have found a candidate and confirmed that its score is preferable
+        // to our current location. See if we have a secondary location in the preferred location already: if not,
+        // then create one.
+        if let Some(replacement) = replacement {
+            // Reject candidate replacements that are not in our preferred AZ: it is not worth the effort
+            // to migrate to a better location when we know there will be some even better location in future (we assume
+            // that if there are no appropriate locations in our preferred AZ now, then there will be eventually).
+            if self.preferred_az_id.is_some()
+                && scheduler.get_node_az(&replacement) != self.preferred_az_id
+            {
```

Comment:
nit: You could assert on this instead. `az_match` is the first member of `NodeAttachmentSchedulingScore`, so `find_better_location` should never return a location that's worse from the AZ pov.

Reply:
> find_better_location should never return a location that's worse from the AZ pov.

That's true, but it might return a location that's equally bad as our current location, if we're already outside the preferred AZ.  What I'm saying with this condition is "if there's a location that's better, but _still_ not in my preferred AZ, then don't spend the resources on migrating: hold out until there is a location in the preferred AZ".

I've rewritten the comment to make that clearer.

Reply:
25ae0ad08fc25193606615a30b597922d814bed0